Data wrangling

#01. Importing libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

#02. Importing data frames

In [2]:
#Importing order data frame

In [3]:
path = r'/Users/Cel/Documents/Data Analytics/09-2023 Instacart Basket Analysis'

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'))

In [5]:
#Importing product data frame
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

#03. Setting up order data frame

In [6]:
#Drop 'eval_set' column
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [7]:
df_ords = df_ords.drop(columns = ['eval_set'])

In [9]:
#Renaming 'order_dow' column
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [10]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#04. Changing data types

In [35]:
#Change 'order_id' from integer to object
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [12]:
df_ords['order_id'].dtype

dtype('O')

#05. Importing departments data frame

In [13]:
#Importing departments data frame
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'))

#06. Setting up departments data frame

In [14]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [15]:
#Transpose df_dep and saved as new name
df_dep_t = df_dep.T

In [16]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
#Adding index column
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [37]:
#Assign row 0 as new header
new_header = df_dep_t.iloc[0]

In [19]:
new_header

0    department
Name: department_id, dtype: object

In [38]:
#Create new df with all rows => 1
df_dep_t_new = df_dep_t[1:]

In [21]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [39]:
#Assign proper header name to new df
df_dep_t_new.columns = new_header

In [40]:
#Department data frame now has proper format
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

#07. Create data dictionary for departments data frame

In [41]:
#Transform df into new dictionary format and saved in new variable 'data_dict'
data_dict = df_dep_t_new.to_dict('index')

In [26]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [42]:
#Use data dictionary to explore department_id in product df
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [44]:
#What does department_id = 19 refer to?
data_dict.get('19')

{'department': 'snacks'}

#08. Subsetting

In [30]:
#indexing department_id=19 column omits rows with other department ids
df_snacks =  df_prods[df_prods['department_id']==19]

In [45]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [46]:
df_prods['department_id']==19

0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

In [32]:
#Another way to create a df with only department_id=19 is to use loc function
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [33]:
df_snacks_2.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

Exercise 4.4

#01. Change variable data type

In [51]:
#Change user_id date type from integer to string
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [52]:
df_ords.head()

order_id user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0  2539329       1                 1                   2                  8   
1  2398795       1                 2                   3                  7   
2   473747       1                 3                   3                 12   
3  2254736       1                 4                   4                  7   
4   431534       1                 5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#02. Change variable name

In [50]:
#Change 'order_number' to 'number_of_orders' without overwriting db
df_ords.rename(columns = {'order_number' : 'number_of_orders'}, inplace = True)

#03. Find busiest ordering hour

In [63]:
#Use value counts to determine most frequent order hour
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

Most frequent order hour is 10am.

#04. Determine meaning of department_id = 4 in df_prods dataframe

In [67]:
data_dict.get('4')

{'department': 'produce'}

A department ID of 4 refers to produce.

#05. Subset for breakfast item sales

In [69]:
#Find department_id for breakfast (14)
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [70]:
#Create subset for all department_id = 14
df_breakfast =  df_prods[df_prods['department_id']==14]

In [73]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#06. Create dinner party subset

In [81]:
#Find department ids for alcohol, deli, beverages, meat/seafood using data dictionary
#Create subset using loc function in a list
df_dinner =  df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [82]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

The dinner data frame has 7650 rows.

#07 Extract data from a specific user

In [96]:
#Data from user_id = 1 is in orders data frame
df_ords_user1 = df_ords[df_ords['user_id']=='1']

In [98]:
#Subset for user_id = 1
df_ords_user1

order_id user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329       1                 1                   2                  8   
1   2398795       1                 2                   3                  7   
2    473747       1                 3                   3                 12   
3   2254736       1                 4                   4                  7   
4    431534       1                 5                   4                 15   
5   3367565       1                 6                   2                  7   
6    550135       1                 7                   1                  9   
7   3108588       1                 8                   1                 14   
8   2295261       1                 9                   1                 16   
9   2550362       1                10                   4                  8   
10  1187899       1                11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

This user has placed 11 orders so far. They normally place their orders sometime in the morning between Monday-Thursday. They appear to order every 2-4 weeks.

#08. Exporting data

In [99]:
#Exporting order data frame
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [100]:
#Exporting departments data frame
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))